### ISSUE TITLE
ensureGreen() times out in IdpRestTestCase tests

### ISSUE THREAD

**Issue by jkakavas** [elastic/elasticsearch#54445]
example: https://gradle-enterprise.elastic.co/s/mjzehzprhvqrc/console-log?task=:x-pack:plugin:identity-provider:qa:idp-rest-tests:integTestRunner

```
rg.elasticsearch.xpack.idp.ManageServiceProviderRestIT > testCreateAndDeleteServiceProvider FAILED
    java.net.SocketTimeoutException: 60 000 milliseconds timeout on connection http-outgoing-5 [ACTIVE]
        at __randomizedtesting.SeedInfo.seed([3B3A4454B4129FCA:1D5BE0E74B4A6BC3]:0)
        at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:808)
        at org.elasticsearch.client.RestClient.performRequest(RestClient.java:248)
        at org.elasticsearch.client.RestClient.performRequest(RestClient.java:251)
        at org.elasticsearch.client.RestClient.performRequest(RestClient.java:235)
        at org.elasticsearch.test.rest.ESRestTestCase.ensureHealth(ESRestTestCase.java:969)
        at org.elasticsearch.test.rest.ESRestTestCase.ensureHealth(ESRestTestCase.java:962)
        at org.elasticsearch.test.rest.ESRestTestCase.ensureGreen(ESRestTestCase.java:949)
        at org.elasticsearch.xpack.idp.ManageServiceProviderRestIT.testCreateAndDeleteServiceProvider(ManageServiceProviderRestIT.java:54)

        Caused by:
        java.net.SocketTimeoutException: 60 000 milliseconds timeout on connection http-outgoing-5 [ACTIVE]
            at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.timeout(HttpAsyncRequestExecutor.java:387)
            at org.apache.http.impl.nio.client.InternalIODispatch.onTimeout(InternalIODispatch.java:92)
            at org.apache.http.impl.nio.client.InternalIODispatch.onTimeout(InternalIODispatch.java:39)
            at org.apache.http.impl.nio.reactor.AbstractIODispatch.timeout(AbstractIODispatch.java:175)
            at org.apache.http.impl.nio.reactor.BaseIOReactor.sessionTimedOut(BaseIOReactor.java:261)
            at org.apache.http.impl.nio.reactor.AbstractIOReactor.timeoutCheck(AbstractIOReactor.java:502)
            at org.apache.http.impl.nio.reactor.BaseIOReactor.validate(BaseIOReactor.java:211)
            at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:280)
            at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
            at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591)
            at java.base/java.lang.Thread.run(Thread.java:834)
REPRODUCE WITH: ./gradlew ':x-pack:plugin:identity-provider:qa:idp-rest-tests:integTestRunner' --tests "org.elasticsearch.xpack.idp.ManageServiceProviderRestIT.testCreateAndDeleteServiceProvider" -Dtests.seed=3B3A4454B4129FCA -Dtests.security.manager=true -Dtests.locale=uk-UA -Dtests.timezone=America/Santarem -Dcompiler.java=13
```

I also hit this in https://github.com/elastic/elasticsearch/pull/54215 where there are two test methods and this would always happen only on the second test of the two that would run. I traced it to the fact that the template was deleted after the first test case ran and then we'd fail to re-create it ( because there was no cluster update received to trigger that (?) ) and solve it temporarily with https://github.com/elastic/elasticsearch/pull/54215/files#diff-f7bdad16f322fc2d554401ea074400b5R1132 but this doesn't seem to apply here where we have only one test. 

Another data point is that this is not reproducible locally, while the https://github.com/elastic/elasticsearch/pull/54215 were

**Comment by tvernum** [comment]
This doesn't really make any sense. The line immediately before the failure is reading a document from that index. The index must exist ... perhaps it is yellow?

**Comment by jkakavas** [comment]
> perhaps it is yellow?

That is actually the case ( looking in my troubleshooting notes from last week and #54215 )

**Comment by tvernum** [comment]
Curious - I thought that cluster only had 1 node, and the index settings are to expand  replicas`0-1` so something isn't how I thought it would be.

But even then, why would it take more than 60s to get to green?

**Comment by romseygeek** [comment]
This has just failed on 7.x as well - can we backport the mute?

https://gradle-enterprise.elastic.co/s/gidsxp5rxevea/tests/ypjwgchf3udh6-nee3awz3nsh5q

**Comment by jkakavas** [comment]
messed up labels so we'll need to wait a little more, but I backported the mute to 7.x and 7.7 @romseygeek

**Comment by tvernum** [comment]
A reminder to myself. Our current theory is that the template gets lost/dropped and so the index gets 1 replica, so on a single node cluster it's yellow not green.

But the question is why the template is being lost.

**Comment by tvernum** [comment]
And that's definitely the cause:
```
» [2020-03-31T10:08:55,234][INFO ][o.e.c.m.MetaDataCreateIndexService] [integTest-0] [saml-service-provider-v1] creating index, cause [auto(bulk api)], templates [], shards [1]/[1], mappings []
```
So, it looks like a bug in our code that isn't guaranteeing that the template is installed before we create the index.

**Comment by tvernum** [comment]
Update: I can reproduce this if I rename `ManageServiceProviderRestIT` to `ZManageServiceProviderRestIT` so that it runs after `WildcardServiceProviderRestIT`.

I wonder if something is causing them to run in a different order in CI sometimes, but not when I try to reproduce it.

At least now I have something that I can work on fixing.

**Comment by tvernum** [comment]
I am now 99% convinced that this is due to:

1. The order to test execution being non-determinisitic. It's file system dependent, so it's mostly alphabetical, but not guaranteed to be so.
2. The template being deleted between tests (because it's [not yet listed as a x-pack system template](https://github.com/elastic/elasticsearch/pull/54215/files#diff-f7bdad16f322fc2d554401ea074400b5)).

When `ManageServiceProviderRestIT` runs second there is no template, so the index is created without an special settings, so the replica count is set to 1, so in a single node cluster has yellow health, rather than the expected green.

**Comment by jkakavas** [comment]
Corresponds to what I've been seeing in #54215 and explains why this is solved by retaining the template when wiping stuff between tests. You can even replicate this by duplicating `testCreateAndDeleteServiceProvider` in the `ManageServiceProviderRestIT`, this would fail consistently he second time around. 

Shouldn't the https://github.com/elastic/elasticsearch/blob/3cc2c80e4c5f7cb41905c68b6f8e2ca1d67ba579/x-pack/plugin/identity-provider/qa/idp-rest-tests/src/test/java/org/elasticsearch/xpack/idp/ManageServiceProviderRestIT.java#L53 cause the index template to be installed, even the second time, in https://github.com/elastic/elasticsearch/blob/b7af8524228891c305b8868ff7f748ec10cf724c/x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderIndex.java#L240 ?

**Comment by tvernum** [comment]
As discussed in our sync, we never reset `templateInstalled` if the template is removed. That field is really tracking _have-we-done-the-work-of-installing-the-template-yet_, and wasn't intended to catch the case where an external process deletes the template after we installed it.

**Comment by jkakavas** [comment]
Resolved in #54215
